Add XDP experiments

micheledallerive opened the pull request:
The folder `xdp/` contains the programs to run the pingpong experiment using two XDP approaches: XDP poll (`pp_poll.c`) and AF_XDP socket (`pp_sock.c`).

Both programs automatically mount their own XDP filter, respectively `pingpong.c` and `pingpong_xsk.c`.

The folder `src/` and `common.h` contain utility methods and macros that are required by both programs. More specifically:
- `common.h`: macros and functions used by all programs (`pp_*`) and XDP programs.
- `src/net.h`: utility functions related to the network and communication between the nodes
- `src/persistence.h`: module to persist the pingpong rounds to file. Contains different flags to choose the type of data that is persisted and the output stream.
- `src/utils.h`: utility functions used in the experiment programs, but not in XDP (either because not needed or not allowed by the bpf compilation).
- `src/xdp-loading.h`: small library to load, attach, detach and pin an XDP filter from a C program.

The programs can be built using CMake and run according to their command-line specifications.

micheledallerive committed:
Change default packet size to 1024 (Ethernet MTU is 1500)

micheledallerive committed:
Add remove-and-retry when the XDP filter is already attached.

micheledallerive committed:
Add loop to send custom packets to server's XDP.

micheledallerive committed:
Add user-space timestamps dump using XDP timestamps.

micheledallerive committed:
Isolate program attach and detach logic in new

micheledallerive committed:
Add conditional printing of logs and error depending on DEBUG.

micheledallerive committed:
Improve code quality and structure.

micheledallerive committed:
Remove the number of iterations from global variables.

micheledallerive committed:
Update XDP program to always send packets to userspace.

Instead of registering the timestamps at XDP-level on the server, timestamps are all registered at application-level, using the combo map+user-space polling.

micheledallerive committed:
Add "persistence" module to simply storage of data to file.

Using helper functions to store data makes the code cleaner, more readable and is much easier to extend and modify.

micheledallerive committed:
Isolate send logic into a "sending thread".

In order to make sure packets are set at the given interval, they must be sent independently on the state of the previous pingpong packet.
To do so, there must be a dedicated thread that sends as many packet as needed.

micheledallerive committed:
Add automatic addresses exchange between client and server.

In order to avoid writing the explicit MAC address and IP address of the nodes involved in the communication, an automatic exchange system is implemented.
Just the IP address of the server is required: the client will send a packet with its addresses, and the server will reply with its own addresses.
This way, no hard-coding or configuration is required, as the server IP address is provided as a command-line argument.

micheledallerive committed:
Improve documentation in the program.

micheledallerive committed:
Fix reboot command in RDMA ansible script.

micheledallerive committed:
Add threaded persistence agent.

It would be interesting to analyse whether having a thread that writes information on the disk affects the measurement less than the program itself.
The persistence agent used in the program can be changed by modifying the flag passed to persistence_init.

micheledallerive committed:
Micro-improvements in polling.

micheledallerive committed:
Add buffered map between XDP and userspace for kernel bypass.

Transform the packets XDP map from a single-valued map to a "circular queue"-like map. This way, packets are not immediately overwritten but are buffered in the map, allowing the userspace to read them even if the polling is slower than the reception of packets.

It is clear that thanks to memory mapping of the eBPF map, polling from userspace is still much faster than the reception of packets; however, because of any spike in delay between two-polls (which is possible when syscalls are involved), some packet could be lost. Using a buffer, the packet is not overridden, and the userspace app has time to poll the old and the new packets.

The implementation is lock-free: since it is a single-producer single-consumer queue, there should be no need for locks.
Most of the time, XDP and userspace will be handling two different sections of the same memory location, creating no hazard. Although extremely unlikely, in a completely unsynchronized implementation there is the possibility of corruption of packets in the case of XDP writing the same space in memory that userspace is reading.
To address this issue, a "magic number" was added at the end of the pingpong payload, which helps recognize the integrity of the packet without need of any checksum: before reading a packets from the map, the userspace app verified that the magic number matches; after reading a packet, its memory space in the eBPF map is zeroed. By doing this, the only possible way to read a packet in userspace is for XDP to have completely written the packet, including the magic number which is placed at the end; in the case of a half-written packet, the magic number would not match stopping the userspace from reading it.

micheledallerive committed:
Add packets interval in command-line arguments.

micheledallerive committed:
Add initial AF_XDP program.

The code is adapted from xdp-tools example on AF_XDP.
A lot of the original code was not useful and thus removed.

Currently working but not testable, since XDP forwards only packets with Ethernet type 0x2002, which can only be sent being custom-made by a program, and the current pp_sock does not send packets yet.

micheledallerive committed:
Improve abstraction of sending packet thread by adding custom send function.

Different types of pingpong might need different types of send functions. For example, pp_poll sends the packets using a normal socket, while pp_sock sends the packets using AF_XDP socket, which has different functions and dynamics compared to a normal socket.
In order to allow this customization, while maintaining the abstraction over the sending thread, a custom function can be passed to the send thread which will be executed when a packet needs to be sent. Some custom variables can be passed to the send thread in order to be used in the send function (for example, to pass the socket to be used to sent the packet).

micheledallerive committed:
Add send and timestamping functionalities to AF_XDP pingpong.

The pingpong using AF_XDP now works, with a customizable amount of packets exchanged at a customizable rate. There is a first phase, just like in pp_poll, in which server and client exchange MAC and IP addresses. Then, the client's send thread starts sending packets at regular intervals using the TX ring of the socket. After the ping is ponged, the timestamps are currently printed to the console.

Currently possible improvements:
- Write the timestamps to file using the persistence agent.
- There is a possible improvement in the packet memory management: the packet headers are written only once as a "template" and then modified for each packet to be sent with the ID; before being sent it's copied to a UMEM frame. There must be a copy: I'm confident it's faster to memcpy the template packet to the UMEM frame than re-creating every time the packet headers on the frame. However, a possible optimization would be instead of writing the pingpong payload on the buffer and the copying the buffer to the UMEM frame, just copy the buffer to the UMEM and then just write the payload information on the UMEM frame. This would avoid the double-write of the payload on both the buffer and the UMEM frame. However, I really think it does not matter, since we are talking about 44 bytes :P.

micheledallerive committed:
Make experiment stop when the number of iterations has been reached.

micheledallerive committed:
Remove statistics and isolate "send packet" functionality in a function.

micheledallerive committed:
Add documentation of the functions in pp_sock.

micheledallerive committed:
Add storage of pingpong data in file for AF_XDP.

micheledallerive committed:
Improve persistence flags and add min-max latency storage.

Using flags, the persistence module can be customized to change where data is printed (file or stdout) and what data is printed (all the timestamps, just the payloads with min and max latency, timestamps with buckets)

micheledallerive committed:
Fix inventory.yaml to work with xl170 machines.

dvr0 commented on xdp/pp_poll.c:
```diff
@@ -0,0 +1,353 @@
+ * @param map_ptr the pointer to the map
+ * @param dest_payload the pointer to the payload to be filled
+ * @param last_index the index of the last payload retrieved
+ * @param busy_poll_next if true, the function busy-waits the next index, otherwise round-robin is used
```
What do you mean by "round-robin" here? Does it not just go loop back to the start since we are looking at the last element?

micheledallerive replied:
While writing the reason behind `busy_poll_next` I realized it is not even needed anymore. 
In the current way, there is no need for `busy_poll_next`: xdp keeps track of last insertion place in the map, and puts the new packet after that one; userspace app keeps track of the latest read position, and resumes reading from the next position. They basically follow each other, so if `map[curr_idx]` does not contain a packet, then there must be no other packet in the map.
I am struggling to recall the reason for `busy_poll_next` to be there; I am sure it was required because the last `map_size` packets received by the machine would not be polled, so the function needed to be told to go through the whole array looking for new packet, not just the next position.
I will remove the flag and related code in the next commits.

dvr0 replied:
Sounds good. I'll let you choose whether you want to do that in this PR or in a future one

dvr0 commented on xdp/src/net.c:
```diff
@@ -0,0 +1,336 @@
+    return 0;
+}
+
+int build_base_packet (char *buf, const uint8_t *src_mac, const uint8_t *dest_mac,
```
What is the reason to go without UDP? You don't have checksum without it, which might be useful for broken packets - do you compensate with it with the `magic` field in the PP proto?

micheledallerive replied:
The idea was to reduce as much the non-required information transmitted in the packet. 
I think UDP offers no particular advantage: its checksum is handled only by end-hosts and it is checked automatically inside the net stack; however, since our packets bypass it, it must be checked manually which could add delay at UDP-level.
The `magic` field in pingpong payload cannot provide much safety from corruption, since it's a fixed value; also, no "wrong magic number" handling was implemented.
Honestly, I had not even though about packet corruption, which I guess becomes very important when the experiment is run over longer periods of time. If it is likely to be a problem, I see no problem in adding the UDP header.

dvr0 replied:
Good point, let's keep it this way for now. We can add checksum logic in the PP payload if we want to add error checking in the future.

dvr0 commented on xdp/common.h:
```diff
@@ -1,9 +1,125 @@
+#define INFO_PACKET_SIZE (ETH_ALEN + sizeof (uint32_t))
+
+// Custom ethernet protocol number
+#define ETH_P_PINGPONG 0x2002
```
ethertype is sometimes used for VLANs, we had some issues with it in some experiments where the switch was dropping packets (I don't recall why or for what network configuration). Let's keep it in mind in case something similar happens

dvr0 reviewed (commented):
Looks good and very well structured!

I am having some issues with the XDP socket: it only saves 2 packets. Here are the steps I did:

```sh
# clone && cd det-bypass
$ cd xdp
$ cmake -DDEBUG=1 .
$ make
$ sudo ldconfig

# server
$ ip a
...
5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 9c:dc:71:5c:ef:b1 brd ff:ff:ff:ff:ff:ff
    altname enp3s0f1np1
    inet 10.10.1.2/24 brd 10.10.1.255 scope global ens1f1np1
       valid_lft forever preferred_lft forever
    inet6 fe80::9edc:71ff:fe5c:efb1/64 scope link 
       valid_lft forever preferred_lft forever
$ sudo ./pp_sock ens1f1np1 start 100
bpf_program__unpin: No such file or directory


# client
$ ip a
...
5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 9c:dc:71:5d:97:b1 brd ff:ff:ff:ff:ff:ff
    altname enp3s0f1np1
    inet 10.10.1.1/24 brd 10.10.1.255 scope global ens1f1np1
       valid_lft forever preferred_lft forever
    inet6 fe80::9edc:71ff:fe5d:97b1/64 scope link 
       valid_lft forever preferred_lft forever
$ sudo ./pp_sock ens1f1np1 start 100 1000000 10.10.1.2
$ cat pingpong_xsk.dat
87: 21944220052732 21942654440099 21942654440158 21944220076448 (LATENCY 11828 ns)
14: 21944140055465 21942574452372 21942574452974 21944140116174 (LATENCY 30053 ns)
```

dvr0 committed:
simple deployment playbook

micheledallerive commented:
> Looks good and very well structured!
> 
> I am having some issues with the XDP socket: it only saves 2 packets. Here are the steps I did:
> 
> ```shell
> # clone && cd det-bypass
> $ cd xdp
> $ cmake -DDEBUG=1 .
> $ make
> $ sudo ldconfig
> 
> # server
> $ ip a
> ...
> 5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
>     link/ether 9c:dc:71:5c:ef:b1 brd ff:ff:ff:ff:ff:ff
>     altname enp3s0f1np1
>     inet 10.10.1.2/24 brd 10.10.1.255 scope global ens1f1np1
>        valid_lft forever preferred_lft forever
>     inet6 fe80::9edc:71ff:fe5c:efb1/64 scope link 
>        valid_lft forever preferred_lft forever
> $ sudo ./pp_sock ens1f1np1 start 100
> bpf_program__unpin: No such file or directory
> 
> 
> # client
> $ ip a
> ...
> 5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
>     link/ether 9c:dc:71:5d:97:b1 brd ff:ff:ff:ff:ff:ff
>     altname enp3s0f1np1
>     inet 10.10.1.1/24 brd 10.10.1.255 scope global ens1f1np1
>        valid_lft forever preferred_lft forever
>     inet6 fe80::9edc:71ff:fe5d:97b1/64 scope link 
>        valid_lft forever preferred_lft forever
> $ sudo ./pp_sock ens1f1np1 start 100 1000000 10.10.1.2
> $ cat pingpong_xsk.dat
> 87: 21944220052732 21942654440099 21942654440158 21944220076448 (LATENCY 11828 ns)
> 14: 21944140055465 21942574452372 21942574452974 21944140116174 (LATENCY 30053 ns)
> ```

@daviderovell0 That's the "min-max" storage, which saves only fastest and slowest packets. For some reason I pushed the code with it. You can change it in the code, when calling `persistence_init` by just modifying the flag that is passed.

dvr0 commented:
> > Looks good and very well structured!
> > I am having some issues with the XDP socket: it only saves 2 packets. Here are the steps I did:
> > ```shell
> > # clone && cd det-bypass
> > $ cd xdp
> > $ cmake -DDEBUG=1 .
> > $ make
> > $ sudo ldconfig
> > 
> > # server
> > $ ip a
> > ...
> > 5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
> >     link/ether 9c:dc:71:5c:ef:b1 brd ff:ff:ff:ff:ff:ff
> >     altname enp3s0f1np1
> >     inet 10.10.1.2/24 brd 10.10.1.255 scope global ens1f1np1
> >        valid_lft forever preferred_lft forever
> >     inet6 fe80::9edc:71ff:fe5c:efb1/64 scope link 
> >        valid_lft forever preferred_lft forever
> > $ sudo ./pp_sock ens1f1np1 start 100
> > bpf_program__unpin: No such file or directory
> > 
> > 
> > # client
> > $ ip a
> > ...
> > 5: ens1f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
> >     link/ether 9c:dc:71:5d:97:b1 brd ff:ff:ff:ff:ff:ff
> >     altname enp3s0f1np1
> >     inet 10.10.1.1/24 brd 10.10.1.255 scope global ens1f1np1
> >        valid_lft forever preferred_lft forever
> >     inet6 fe80::9edc:71ff:fe5d:97b1/64 scope link 
> >        valid_lft forever preferred_lft forever
> > $ sudo ./pp_sock ens1f1np1 start 100 1000000 10.10.1.2
> > $ cat pingpong_xsk.dat
> > 87: 21944220052732 21942654440099 21942654440158 21944220076448 (LATENCY 11828 ns)
> > 14: 21944140055465 21942574452372 21942574452974 21944140116174 (LATENCY 30053 ns)
> > ```
> 
> @daviderovell0 That's the "min-max" storage, which saves only fastest and slowest packets. For some reason I pushed the code with it. You can change it in the code, when calling `persistence_init` by just modifying the flag that is passed.

Works, didn't see that sorry. Could we perhaps have that flag as an argument to the program?

It can also be a future feature. Feel free to merge without

dvr0 committed:
exclude large files in dummy deploy script

micheledallerive committed:
Update XDP ansible to add ntuple rule for pingpong packets.

micheledallerive committed:
Change user-space poll function to constantly poll only the next index.

In an ideal context, the userspace app would constantly read the packet that XDP writes, staying in syntony.
This commit solves the problems that represents hazard to such balance, precisely:
- Un-synchronized access to the "write index" in XDP program, in case multiple instances are run on different cores. This problem caused gaps in the map that the userspace app had to wait to be filled before going on.
- Overwrite-apocalypse: in cases of enormous delay in the userspace side, the XDP program would manage to write packets in the whole app, ending up finding an old packet and starting to drop new packets. However, in this case, the "write index" in XDP must stay the same, otherwise it will resume writing in an arbitrary point of the map, leaving gaps that will lead to an unexpected behaviour. This problem is NOT solved: this commit will be used as a reference for this error; next commit will un-comment the solution.

micheledallerive committed:
Solve polling error "Overwrite-apocalypse" described in the previous commit.

micheledallerive committed:
Add XDP playbook with recent experiment.

micheledallerive committed:
Disable DUMP_MAP by default.

dvr0 committed:
renaming analysis/lab_book for consistency

dvr0 reviewed (approved):
looks gooood
